### ISSUE TITLE
FEFaceEvaluation cannot be inserted into a std::vector

### ISSUE THREAD

**Issue by johnbcoughlin** [dealii/dealii#17563]
- Using deal.II v9.5.1 on MacOS 14.6.1, compiled with candi

The following example fails to compile:
```
#include <deal.II/matrix_free/fe_evaluation.h>

using namespace dealii;

int main() {
    MatrixFree<1, double> mf;
    std::vector<FEFaceEvaluation<1, -1, 0, 5, double>> phis;
    phis.emplace_back(mf, true, 0, 0, 0);
}
```

The root cause compiler error is found inside `fe_evaluation.h`:
```
/Users/jack/dealii-candi/deal.II-v9.5.1/include/deal.II/matrix_free/fe_evaluation.h:3064:28: error: incompatible pointer types assigning to 'const MappingInfoStorageType *' (aka 'const MappingInfoStorage<(true ? 1 - 1 : 1), 1, dealii::VectorizedArray<double, 1>> *') from 'MappingInfoStorage<1, 1, VectorizedArray<double, 1>> *'
      this->mapping_data = &this->mapped_geometry->get_data_storage();
                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

The same example but with an `FEEvaluation` succeeds, which leads me to suspect this may not be part of the intentional design surface of the MatrixFree infrastructure.

To give an idea of why I want to stick these in a vector: I'm writing a multispecies fluid code, where the number of species isn't known until runtime. Each species contributes 5 components to the FESystem, and for some operations we need to act on or use only the components from one or the other species. The best API design is to provide the user of the library with a vector of `FEFaceEvaluation`, one for each species, so they can ask for the evaluations they need (via `EvaluationFlags`).
In principle it should be possible to limit users to a small number of species known at compile time, but it would be onerous and involve a lot of extra templating that I'd prefer to avoid.

**Comment by kronbichler** [comment]
Indeed, this is supposed to work but it wouldn't because we did not test it for a while. I will open a pull request to fix it, thank you for the hint.

**Comment by bangerth** [comment]
In the meantime (unless you want to switch to the `master` branch outright), you can probably use a vector of unique-pointers.
